### PR TITLE
Cache output for profiles show

### PR DESCRIPTION
**Pull request by grahamgilbert**

### CODE REVIEW

**Comment by erikng** [comment]
What happens if some other tools runs the command? Shouldn't we instead check for exit status or the string when the command is blocked rather than assuming osquery is the only thing polling/attempting to run this command?

**Comment by grahamgilbert** [comment]
We _do_ check for the rate limiting string. Am I missing something here?

**Comment by erikng** [comment]
I think what's unclear and it could be me misreading the code, is what happens when there is no json and the device is already rate limited?

I assume no data will be returned, but then it will wait 24 hours to try again. But if another tool runs again, you could perpetually be waiting.

Wonder if there is somewhere private that we could obtain the last check data, that way we are more likely to succeed in getting the data.

**Comment by grahamgilbert** [comment]
Maybe, I’m open to ideas on that one. At the moment this is the best way I can see this succeeding.

File: tables/mdm/mdm.go [hunk 1]
```diff
@@ -65,22 +77,17 @@ func MDMInfoColumns() []table.ColumnDefinition {
 }
 
 func MDMInfoGenerate(ctx context.Context, queryContext table.QueryContext) ([]map[string]string, error) {
-	profiles, err := getMDMProfile()
-	if err != nil {
-		return nil, err
-	}
+	profiles, _ := getMDMProfile()
```
Comment (discentem) [inline tables/mdm/mdm.go:80]: Why did you stop handling the err and instead throw it out? Please leave a comment to justify.

File: tables/mdm/mdm.go [hunk 2]
```diff
@@ -65,22 +77,17 @@ func MDMInfoColumns() []table.ColumnDefinition {
...
 	if err == nil { // only supported on 10.13.4+
 		depEnrolled = strconv.FormatBool(status.DEPEnrolled)
 		userApproved = strconv.FormatBool(status.UserApproved)
 	}
 
-	depstatus, err := getDEPStatus()
-	if err == nil {
-		depCapable = strconv.FormatBool(depstatus.DEPCapable)
-	}
+	depstatus, _ := getDEPStatus(status)
```
Comment (discentem) [inline tables/mdm/mdm.go:89]: Same comment as above. Please leave a code comment explaining why its safe to throw out the error

File: tables/mdm/mdm.go [hunk 3]
```diff
@@ -160,20 +167,103 @@ func getMDMProfileStatus() (profileStatus, error) {
...
+				depCapable := getCachedDEPStatus()
+				depstatus.DEPCapable = depCapable
+				return depstatus, nil
+			}
+			return depstatus, nil
+		}
+
+		lines := bytes.Split(out, []byte("\n"))
+
+		if len(lines) > 3 {
```
Comment (discentem) [inline tables/mdm/mdm.go:192]: optional: Please add a comment explaining the significance of `len(lines) > 3` here.

File: tables/mdm/mdm.go [hunk 4]
```diff
@@ -160,20 +167,103 @@ func getMDMProfileStatus() (profileStatus, error) {
...
+	plistFile, err := os.Open(CloudConfigTimerCheck)
 	if err != nil {
-		fmt.Println(err)
+		// could not open file
+		return false
 	}
 
-	lines := bytes.Split(out, []byte("\n"))
+	byteValue, err := ioutil.ReadAll(plistFile)
+	if err != nil {
```
Comment (discentem) [inline tables/mdm/mdm.go:217]: It seems bad on the surface to blackhole these errors. We should return them I think?

File: tables/mdm/mdm.go [hunk 5]
```diff
@@ -160,20 +167,103 @@ func getMDMProfileStatus() (profileStatus, error) {
...
+		// could not read file to bytes
+		return false
+	}
+
+	err = plist.Unmarshal(byteValue, &cloudConfigTimerCheck)
+	if err != nil {
+		// could not unmarshal file from bytes to cloudConfigTimerCheck type
+		return false
```
Comment (discentem) [inline tables/mdm/mdm.go:225]: same comment as above

File: tables/mdm/mdm.go [hunk 6]
```diff
@@ -43,9 +46,18 @@ type profileStatus struct {
 }
 
 type depStatus struct {
-	DEPCapable bool
+	DEPCapable  bool `json:"dep_capable"`
```
Comment (discentem) [inline tables/mdm/mdm.go:49]: optional: Please consider adding comments giving context for what these struct fields represent.

File: tables/mdm/mdm.go [hunk 7]
```diff
@@ -43,9 +46,18 @@ type profileStatus struct {
...
+	RateLimited bool `json:"rate_limited"`
```
Comment (discentem) [inline tables/mdm/mdm.go:50]: field names should technically not be exported/capitalized unless they are meant to be exported but the struct itself is not exported. Is this for unmarshaling?

File: tables/mdm/mdm.go (hunk 4 above)
Comment (grahamgilbert) [reply tables/mdm/mdm.go:217]: These being errors aren't actually errors - i.e. the program isn't failing, they're just representing a valid state. Also, osquery extensions don't generally log errors, they just do not return anything if there is an issue or they aren't applicable.

**Review by discentem — APPROVED** [review]

**Pull request merged**